accepted/0027-remove-depth-outdated.md

```diff
@@ -0,0 +1,89 @@
+
+In practice, what most users care about is `--depth 9999` which should display a list of _all_ outdated dependencies on the tree. Also, npm v7 has already recently removed the `--depth` flag from `npm update` in favor of updating all nodes in the tree at every depth, which defeats the purpose of having the functionality to inspect outdated packages at specific depths if they are all going to be updated either way.
+
+Removing `--depth` in favor of a new option flag named `--all` that will display all outdated dependencies on the tree makes for a clearer and more obvious output.
+
+## Detailed Explanation
+Going any level deep further than the top-level deps opens a complex scenario where a single dependency might need different versions to satisfy requirements from different packages depending on it. Basically, what this means is that the "wanted" version of a nested dependency could be either a  _single_ semver range or a _set_ of semver ranges. 
+
+When `--depth` is used on v6, the same dependency and its "wanted" value will be displayed and repeated on console with their location as the sole way of differentiating its level of nesting. This current output can tend to clog the console since the command doesn't tell appart deduped dependencies and we end up with a lot of noise that could be summarized differently.
+
+Ex.
+```
+➜  arborist git:(master) npm outdated --depth 9999 tar
+Package  Current  Wanted  Latest  Location
+tar       4.4.13  4.4.13   6.0.2  @npmcli/arborist > @npmcli/run-script > node-gyp
+tar        6.0.1   6.0.2   6.0.2  @npmcli/arborist > pacote > cacache
+tar        6.0.1   6.0.2   6.0.2  @npmcli/arborist > pacote > npm-registry-fetch > make-fetch-happen > cacache
+tar        6.0.1   6.0.2   6.0.2  @npmcli/arborist > pacote
+
+➜  arborist git:(master) npm ls tar
+@npmcli/arborist@0.0.0-pre.11 /Users/claudiahdz/npm/arborist
+├─┬ @npmcli/run-script@1.2.1
+│ └─┬ node-gyp@6.1.0
+│   └── tar@4.4.13
+└─┬ pacote@11.1.0
+  ├─┬ cacache@15.0.0
+  │ └── tar@6.0.1  deduped
+  └── tar@6.0.1
+```
+
+In this example, we only care about two versions of `tar`, `4.4.13` and `6.0.2`. Instead we are shown 4 lines of output.
+
+Moreover, npm v6 currently has inconsistent behavior. Doing `npm outdated --depth 9999` displays all outdated packages that are direct children of the root package, missing out dependencies that are nested at other levels. This gives a wrong impression since we are not really displaying _all_ outdated dependencies of the tree. However, doing `npm outdated foo --depth 9999`, will indeed display all appearances of `foo` on the tree no matter if they are direct children of the root package or not. This is precisely the functionality that displaying all deps should have too (`npm outdated` with no flags). 
```

Comment:
> Doing `npm outdated --depth 9999` displays all outdated packages that are direct children of the root package, missing out dependencies that are nested at other levels. This gives a wrong impression since we are not really displaying _all_ outdated dependencies of the tree.

I don't think that's the case, unless I'm misreading this maybe?  `npm outdated --depth=9999` includes this output, for example:

```
log-update                  3.4.0    3.4.0    4.0.0  npm-audit-report > tap > ink
log-update                  3.4.0    3.4.0    4.0.0  npm-audit-report > tap > treport > ink
```

Reply:
I think in that case that particular dependency is being deduped and is _indeed_ a direct child of your root package, dependencies that are not deduped and are nested within their own dependant, do not seem to show up when doing only `npm oudated --depth 9999` but DO show up when you do `npm oudated --depth 9999 <packageName>`.

Reply:
Hm, it's not though.  `log-update` is a dependency of ink, and my package depends on tap, which depends on ink.

```
$ find node_modules -name log-update
node_modules/tap/node_modules/log-update
```

It is a good example of the fact that `--depth` is pretty weird when talking about a dependency graph, though.  It looks like I have two outdated versions of `log-update`, when really they're the same dep node.

Reply:
Yeah, correct. The thing is that `log-update` is a direct child of `tap`.  So I misphrased that by saying the command shows only direct children of the _root_ package when I meant to say that the v6 algorithm is iterating recursively over the node's direct children (on the file system) instead of over `node.edgesOut`. 

Try doing `npm outdated --depth 9999 ansi-escapes` (basically any specific dep inside `node_modules/tap/log-update/node_modules`) and you'll see they are not displayed when calling outdated without a package-name.

Using `tree.inventory` and `node.edgesIn` on v7 will definitely solve this problem and properly give us a complete list of all outdated packages.